### PR TITLE
Lthread

### PR DESCRIPTION
**Pull request by vloginova**
- Improve lthread interface: add lthread_(yield/resume) support
- Make priority functions actually common for threads and lthreads
- Fix some significant bugs in lthread waiting API
- Add demo application game Race
- Add templates for x86 and stm32vl with lthread-only scheduler.

### CODE REVIEW

File: src/kernel/sched/idle_light.c [hunk 1]
```diff
@@ -1,29 +1,33 @@
...
-	lt = lthread_create(idle_run, NULL);
-	if (0 != err(lt)) {
-		return err(lt);
-	}
+	lthread_init(&idle, idle_run);
 
-	sched_change_priority(&lt->schedee, SCHED_PRIORITY_MIN);
-#if 0 //XXX
+	schedee_priority_set(&idle.schedee, SCHED_PRIORITY_MIN);
+#if 0 //XX
```
Comment (anton-bondarev) [inline src/kernel/sched/idle_light.c]: Does it need for SMP? If yes may be it's better to use macro for this SMP_USE MULTICORE_SUPPORT

File: src/include/kernel/sched/schedee_priority.h [hunk 2]
```diff
@@ -1,26 +1,63 @@
 /**
  * @file
- * @brief
+ * @brief TODO documentation for sched_priority.h -- Eldar Abusalimov
  *
- * @date 21.11.2013
- * @author Andrey Kokorev
+ * @date Sep 6, 2011
+ * @author Eldar Abusalimov
```
Comment (anton-bondarev) [inline src/include/kernel/sched/schedee_priority.h]: You should add youself to authors

File: src/kernel/sched/timing/none.h [hunk 3]
```diff
@@ -11,6 +11,10 @@
 #include <sys/cdefs.h>
 #include <sys/types.h>
 
+#include <kernel/sched/sched_timing.h>
+
+struct schedee;
```
Comment (anton-bondarev) [inline src/kernel/sched/timing/none.h]: Why does this declaration need?

File: src/kernel/lthread/lthread.c [hunk 4]
```diff
@@ -39,37 +37,47 @@ static struct schedee *lthread_process(struct schedee *prev,
...
+	struct schedee *self = schedee_get_current();
 
-	if (!run) {
-		return err_ptr(EINVAL);
-	}
+	assert(lt);
 
-	if (!(lt = (struct lthread *) pool_alloc(&lthread_pool))) {
-		return err_ptr(ENOMEM);
+	sched_wait_prepare();
```
Comment (abusalimov) [inline src/kernel/lthread/lthread.c]: Shouldn't this be done through regular `SCHED_WAIT` macro?

File: src/kernel/lthread/lthread.h [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+ * @p target_label.
+ * @note
+ *   This function should be used in pair with return statement.
+ * @param initial_label
+ *   The beginning label.
+ * @param target_label
+ *   Label from which lthread should be resumed.
+ */
+#define lthread_yield(initial_label, target_label) \
+			(target_label) - (initial_label);
```
Comment (abusalimov) [inline src/kernel/lthread/lthread.h]: There must be no trailing semicolon, and the whole expression must be enclosed with parens.

File: src/kernel/lthread/lthread.h [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+ * Returns label from which run function should start with.
+ * @note
+ *   This function should be used in pair with goto statement.
+ * @param lt
+ *   The current light thread.
+ * @param initial_label
+ *   The beginning label.
+ */
+#define lthread_resume(lt, initial_label) \
+			*((lt)->label_offset + (initial_label))
```
Comment (abusalimov) [inline src/kernel/lthread/lthread.h]: Not sure, but this probably has to be inside parens too.

Comment (abusalimov) [inline src/kernel/lthread/lthread.h]: Why are these macros at the first place? Inline functions would be better here, IMO.

**Comment by anton-bondarev** [comment]
This pull request if too big to review. In future it will be better to use short branch. For this task may be split lthread adding, switching to lthread net, timers, etc., desing race demo game.

File: src/kernel/lthread/lthread.c (hunk 4 above)
Comment (vloginova) [inline src/kernel/lthread/lthread.c]: This macro uses the function from https://github.com/embox/embox/blob/master/src/kernel/thread/thread_sched_wait.c file that depends on some thread-specific headers and functions.

File: src/kernel/sched/timing/none.h (hunk 3 above)
Comment (vloginova) [inline src/kernel/sched/timing/none.h]: Because functions declared in the file have ```struct schedee *``` as an argument.

**Comment by AntonKozlov** [comment]
arm/nxt_bt_pnet:
```
arm-none-eabi-gcc  -pipe --debug-prefix-map=`pwd`= --debug-prefix-map=./= -fno-strict-aliasing -fno-common -Wall -Werror -Wundef -Wno-trigraphs -Wno-char-subscripts -Wformat -std=gnu99 -O0 -g -march=armv4 -mtune=arm7tdmi -U__linux__ -Ulinux -U__linux -D__EMBOX__ -D__unix -D"__impl_x(path)=<../path>" -imacros build/base/gen/config.h -I./build/include -I./src/include -I./src/arch/arm/include -Ibuild/base/gen/include -Ibuild/base/gen/src/include -Ibuild/base/gen/src/include -Ibuild/base/gen/src/arch/arm/include -I./platform/lego_nxt/include -I./src/compat/linux/include -I./src/compat/posix/include -I./src/compat/libc/include -nostdinc -MMD -MP  -include build/base/gen/include/config/embox/pnet/net_entry.h -D__EMBUILD_MOD__=embox__pnet__net_entry -c -o build/base/obj/src/pnet/net_entry.o src/pnet/net_entry.c
src/pnet/net_entry.c: In function ‘pnet_rx_action’:
src/pnet/net_entry.c:76:2: error: return makes integer from pointer without a cast [-Werror]
  return NULL;
  ^
cc1: all warnings being treated as errors
```
x86/smp:
```
src/kernel/sched/strategy/runq/list_array.c: In function ‘runq_insert’:
src/kernel/sched/strategy/runq/list_array.c:35:4: error: passing argument 1 of ‘schedee_priority_get’ from incompatible pointer type [-Werror]
    &queue->list[schedee_priority_get(&schedee->priority)]);
    ^
In file included from ./src/include/kernel/sched.h:21:0,
                 from ./src/include/kernel/thread/types.h:22,
                 from ./src/include/kernel/thread.h:18,
                 from ./build/include/../../src/kernel/task/multi.h:17,
                 from build/base/gen/include/module/embox/kernel/task/multi.h:10,
                 from build/base/gen/include/module/embox/kernel/task/api.h:7,
                 from ./src/include/kernel/task.h:17,
                 from src/kernel/sched/strategy/runq/list_array.c:11:
./src/include/kernel/sched/schedee_priority.h:46:12: note: expected ‘struct schedee *’ but argument is of type ‘struct schedee_priority *’
 extern int schedee_priority_get(struct schedee *schedee);
            ^
cc1: all warnings being treated as errors
make[4]: *** [build/base/obj/src/kernel/sched/strategy/runq/list_array.o] Error 1
```

**Pull request merged**